🆕 Version `0.1.0-beta`

codermarcos opened the pull request:
# 📝 Description

This Pull Request make a lot of things!

- Changes the package name from `git-message-ai` to `git-ai` since the package not only generates messages
- Allow omit ticket provider and ticket id by this way the AI will read only the DIFF
- Allow send more than one ticket
- Change JIRA authentication to use USER and API KEY
- Refactor prompt
- Start the code to create a pull request into Github

### 🗃️ Changes

On branch:

> **feature/version-2**


Changes to be committed:

> deleted:    git-message.json
 modified:   .gitignore
 modified:   Cargo.lock
 modified:   README.md
 modified:   src/ai/mod.rs
 modified:   src/ai/open_ai.rs
 modified:   src/cli/direct.rs
 modified:   src/cli/interactive.rs
 modified:   src/configuration.rs
 modified:   src/git.rs
 modified:   src/main.rs
 modified:   src/vcs/github.rs

codermarcos committed:
🔄 Sync code between machines
On branch temporary
Changes to be committed:
	modified:   Cargo.lock
	modified:   Cargo.toml
	modified:   README.md
	new file:   docs/NODE.md
	new file:   docs/README.md
	new file:   git-message.json
	modified:   src/ai/mod.rs
	modified:   src/ai/open_ai.rs
	modified:   src/cli/direct.rs
	modified:   src/cli/interactive.rs
	new file:   src/configuration.rs
	modified:   src/git.rs
	modified:   src/main.rs
	modified:   src/providers/jira.rs
	modified:   src/providers/mod.rs
	new file:   src/vcs/github.rs
	new file:   src/vcs/mod.rs

codermarcos committed:
🆕 add feature to dialoguer

codermarcos committed:
🔀 Merge branch 'main' of github.com:codermarcos/git-message-ai into temporary

Conflicts:
	src/ai/mod.rs
	src/ai/open_ai.rs
	src/cli/direct.rs
	src/git.rs
	src/main.rs

All conflicts fixed but you are still merging.
Changes to be committed:
	modified:   .gitignore
	modified:   src/ai/mod.rs
	modified:   src/ai/open_ai.rs
	modified:   src/cli/direct.rs
	modified:   src/configuration.rs
	modified:   src/main.rs
	modified:   src/providers/jira.rs

codermarcos committed:
🆕 add base64 converter and remove dialoguer history
On branch temporary
Changes to be committed:
	modified:   Cargo.lock
	modified:   Cargo.toml

codermarcos committed:
📚 update documentation
On branch temporary
Changes to be committed:
	modified:   README.md

codermarcos committed:
🆕 add new git commands `diff`, `log`, and `branch`
On branch temporary
Changes to be committed:
	modified:   src/git.rs

codermarcos committed:
🆕 add necessary fields to create a pr to github
On branch temporary
Changes to be committed:
	modified:   src/vcs/github.rs

codermarcos committed:
🆕 make some configurations optional
On branch temporary
Changes to be committed:
	modified:   src/configuration.rs

codermarcos committed:
🆕 add command to create pr
On branch temporary
Changes to be committed:
	modified:   src/cli/direct.rs

codermarcos committed:
🆕 allow multiple tasks or any task
On branch temporary
Changes to be committed:
	modified:   src/ai/mod.rs
	modified:   src/ai/open_ai.rs
	modified:   src/cli/interactive.rs
	modified:   src/cli/mod.rs
	modified:   src/main.rs
	modified:   src/providers/jira.rs
	modified:   src/providers/mod.rs

codermarcos committed:
🆕 improve ai prompt
On branch temporary
Changes to be committed:
	modified:   src/ai/open_ai.rs

codermarcos committed:
🆕 add feature to create a pull request for Github
On branch temporary
Changes to be committed:
	modified:   src/ai/mod.rs
	modified:   src/ai/open_ai.rs
	modified:   src/cli/interactive.rs
	modified:   src/main.rs
	modified:   src/vcs/mod.rs

codermarcos committed:
🐛 fix problem at configuration
On branch temporary
Changes to be committed:
	modified:   src/configuration.rs

codermarcos committed:
🆕 add option to dry run
On branch temporary
Changes to be committed:
	modified:   src/cli/direct.rs

codermarcos committed:
🆕 add options to work with git hooks
On branch temporary
Changes to be committed:
	modified:   src/git.rs

codermarcos committed:
🎨 refactor to use the same name to variables
On branch temporary
Changes to be committed:
	modified:   src/ai/mod.rs

codermarcos committed:
🆕 add option to dry run when generate messages
On branch temporary
Changes to be committed:
	modified:   src/main.rs

codermarcos committed:
🆕 add command to run add hook
On branch temporary
Changes to be committed:
	modified:   src/main.rs

codermarcos committed:
🐛 create a fn for each repeated params
On branch temporary
Changes to be committed:
	modified:   src/cli/direct.rs

codermarcos committed:
🔥 remove configuration file
On branch temporary
Changes to be committed:
	deleted:    git-message.json

codermarcos committed:
🆕 add option to install githook
On branch temporary
Changes to be committed:
	modified:   src/cli/interactive.rs

codermarcos committed:
🐛 fix unfinished code adding _
On branch temporary
Changes to be committed:
	modified:   src/main.rs

codermarcos commented on src/cli/direct.rs:
```diff
@@ -1,42 +1,87 @@
 pub fn run() -> Command {
-    Command::new("git-message-ai")
-        .about("A CLI to generate your git messages!")
+    let dry_run_arg = || {
```
It `||` is like an arrow function but the correct name is [closures](https://doc.rust-lang.org/rust-by-example/fn/closures.html)!
I'm using it because we are using this params in more than one part so it is to avoid repeated code!

codermarcos commented on README.md:
```diff
@@ -13,44 +15,60 @@
-GMA_JIRA_AUTHORIZATION=<jira authorization can be your user and password as a basic>
-GMA_OPENAI_TOKEN=<you can get it here: https://platform.openai.com/api-keys>
+
+# If you are using JIRA
+GA_JIRA_API_KEY=<you can get it here: https://id.atlassian.com/manage-profile/security/api-tokens>
+GA_JIRA_USER=<your user to sign in into jira>
```
Now the prefix is `GA` instead `GMA` because the package name we are receiving `GA_JIRA_API_KEY` and `GA_JIRA_USER`!

edgardleal replied:
What about use [this one](https://docs.rs/config/latest/config/)?

codermarcos commented on src/ai/mod.rs:
```diff
@@ -20,23 +14,46 @@ pub const AVAILABLE_AI: &[&str; 1] = &["OpenAI"];
 pub async fn generate_message(
     configuration: super::Configuration,
-    ticket: Option<Ticket>,
+    tickets: Vec<Ticket>,
```
Now for the ticket we are receiving a vector so the user can pass nothing, one or more tickets

codermarcos commented on src/ai/open_ai.rs:
```diff
@@ -40,70 +40,142 @@ pub async fn generate_message(
+    }
+}
+
+pub fn get_prompt_for_commit(
```
I've created a function to get the prompt since we can use a different prompt to create pull requests

codermarcos commented on src/ai/open_ai.rs:
```diff
@@ -40,70 +40,142 @@ pub async fn generate_message(
         ChatMessage {
-            role: String::from("system"),
+            role: "system",
             content: String::from(
-                r#"You should act as a senios software developer that will generate git commit messages following
-the semantic commit convention to generate beautiful commit message that if easy to read.
-Wenever you receive a Ticket information, you should use these information to help contextualize the
-updates you will receive on git diff.
-You should use emojis whenever it is usefull to help the developers understand what was implemented in the git diff.
-You should take care of not violate the semantic commit rules when use emojis.
-You should not include any extra content on your response, only the final commit message ready to be included on a git commit.
-  "#,
+                "You should act as a senior software developer \
+                that will generate beautiful git commit messages following the provided \
+                instructions that would be easy to read. Whenever you receive a Ticket \
+                information, you should use these information to help contextualize the \
+                updates you will receive on git diff. You should follow the provided \
+                instructions to create a message that matches with the user pattern. \
+                You should take care of not violate the provided instructions. You \
+                should not include any extra content on your response, only the final \
+                commit message ready to be included on a git commit.",
             ),
         },
         ChatMessage {
-            role: "user".to_string(),
+            role: "user",
```
Change from `String` to `str`

codermarcos commented on src/ai/open_ai.rs:
```diff
@@ -40,70 +40,142 @@ pub async fn generate_message(
-        }),
-        None => {}
+
+    if tickets.len() != 0 {
+        messages.push(ChatMessage {
+            role: "user",
+            content: tickets
+                .iter()
+                .map(|ticket| {
+                    format!(
+                        "Task id: {}\nTask title: {}\nTask description: {}\n",
+                        ticket.id, ticket.title, ticket.description,
+                    )
+                })
+                .collect::<Vec<String>>()
+                .join("\n---\n"),
+        });
     }
```
If there are one or more tickets

codermarcos commented on src/ai/open_ai.rs:
```diff
@@ -40,70 +40,142 @@ pub async fn generate_message(
-        }),
-        None => {}
+
+    if tickets.len() != 0 {
+        messages.push(ChatMessage {
+            role: "user",
+            content: tickets
+                .iter()
+                .map(|ticket| {
+                    format!(
+                        "Task id: {}\nTask title: {}\nTask description: {}\n",
+                        ticket.id, ticket.title, ticket.description,
+                    )
+                })
+                .collect::<Vec<String>>()
+                .join("\n---\n"),
+        });
     }
-    let json_body = serde_json::json!({
-        "model": model,
-        "temperature": 0.2,
-        "messages": messages
-    });
 
-    let response = client
-        .post(url)
-        .headers(headers)
-        .body(serde_json::to_string(&json_body).unwrap())
-        .send()
-        .await;
+    serde_json::json!(messages)
+}
```
The diff is a little bit stranger here but now the function `generate_message` only receives a `serde::Value` which is the massages that we are sending to open ai and we have two new functions to create this `serde::Value` which is `get_prompt_for_pull_request` and `get_prompt_for_commit`!

codermarcos commented on src/configuration.rs:
```diff
@@ -0,0 +1,68 @@
+        );
+    }
+
+    pub fn read() -> Self {
+        let configuration_file: &Path = Path::new(CONFIGURATION_FILE);
+
+        if !configuration_file.exists() {
+            println!("⚠️ Configuration file `git-message.json` not found using default!");
+            return Self::new();
+        }
+
+        let file = fs::read_to_string(configuration_file)
+            .expect(format!("Failed to read configuration file `{}`", CONFIGURATION_FILE).as_str());
+
+        serde_json::from_str(&file).unwrap()
+    }
```
Make the configuration file optional!

codermarcos commented on src/main.rs:
```diff
@@ -43,36 +17,72 @@ async fn main() -> Result<(), Error> {
+            let commits = git::get_different_commits_between_branches(&head, &base).await;
+
+
+            let _message =
+                ai::generate_pull_request(&configuration, tickets.unwrap(), commits.unwrap())
+                    .await?;
+
+            let (title, body) =
+                cli::interactive::get_message_change(&String::from(""), &String::from(""));
```
Unfinished code!

codermarcos commented on src/providers/jira.rs:
```diff
@@ -37,36 +37,70 @@ pub async fn get_ticket(
+) -> Result<Vec<Value>, Error> {
+    let (client, headers) = get_client();
+
+    let bodies = future::join_all(ticket_ids.iter().map(|id| {
```
future is like a Promise, here is like Promise.all!

codermarcos commented on src/vcs/mod.rs:
Creates a module to communicate with the chosed vcs which for now is just the Github!

codermarcos reviewed (commented):
Some important hints

codermarcos committed:
📚 update documentation
On branch feature/version-2
Changes to be committed:
	modified:   README.md
	modified:   docs/NODE.md

edgardleal reviewed (approved):
Looks Good to Me

codermarcos committed:
🆕 add file to retrive environment variables
On branch feature/version-2
Changes to be committed:
	new file:   src/secrets.rs

codermarcos committed:
🎨 change configuration file name
On branch feature/version-2
Changes to be committed:
	modified:   src/configuration.rs

codermarcos committed:
🎨 use new mod to read environment variables
On branch feature/version-2
Changes to be committed:
	modified:   src/ai/open_ai.rs
	modified:   src/main.rs
	modified:   src/providers/jira.rs
	modified:   src/vcs/github.rs

codermarcos committed:
🆕 add new packages
On branch feature/version-2
Changes to be committed:
	modified:   Cargo.lock
	modified:   Cargo.toml

codermarcos committed:
Merge branch 'main' of github.com:codermarcos/git-message-ai into feature/version-2

Conflicts:
	src/ai/mod.rs
	src/ai/open_ai.rs

On branch feature/version-2
All conflicts fixed but you are still merging.
Changes to be committed:
	modified:   src/ai/mod.rs
	modified:   src/ai/open_ai.rs
	modified:   src/cli/interactive.rs
